### PR TITLE
[SPARK-9704] [ML] Made ProbabilisticClassifier, Identifiable, VectorUDT public APIs

### PR DESCRIPTION
**Pull request by jkbradley**
Made ProbabilisticClassifier, Identifiable, VectorUDT public.  All are annotated as DeveloperApi.

CC: @mengxr @EronWright

### CODE REVIEW

**Comment by EronWright** [comment]
+1

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 1]
```diff
@@ -164,10 +164,11 @@ sealed trait Vector extends Serializable {
  * User-defined type for [[Vector]] which allows easy interaction with SQL
  * via [[org.apache.spark.sql.DataFrame]].
  *
- * NOTE: This is currently private[spark] but will be made public later once it is stabilized.
+ * WARNING: The [[UserDefinedType]] has NOT been stabilized, so users should be aware that this API
+ *          may need to change in future releases.
  */
 @DeveloperApi
-private[spark] class VectorUDT extends UserDefinedType[Vector] {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Shall we mark it `@AlphaComponent` since it is very likely to change?

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: ok

**Comment by jkbradley** [comment]
MIMA's acting weird locally, so I'll rely on Jenkins to test it.

**Comment by jkbradley** [comment]
I'll test this once more and then merge it.

**Comment by jkbradley** [comment]
jenkins test this please

**Comment by jkbradley** [comment]
jenkins test this please

**Comment by mengxr** [comment]
test this please

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 2]
```diff
@@ -205,7 +205,11 @@ private[spark] abstract class ProbabilisticClassificationModel[
   }
 }
 
-private[ml] object ProbabilisticClassificationModel {
+/**
+ * :: DeveloperApi ::
+ */
+@DeveloperApi
+object ProbabilisticClassificationModel {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: Do we want to make it public? It exposes the `normalizeToProbabilitiesInPlace` method.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 3]
```diff
@@ -159,15 +159,16 @@ sealed trait Vector extends Serializable {
...
 
 /**
- * :: DeveloperApi ::
+ * :: AlphaComponent ::
  *
  * User-defined type for [[Vector]] which allows easy interaction with SQL
  * via [[org.apache.spark.sql.DataFrame]].
  *
- * NOTE: This is currently private[spark] but will be made public later once it is stabilized.
+ * WARNING: The [[UserDefinedType]] has NOT been stabilized, so users should be aware that this API
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: The warning message is not necessary. This is what `AlphaComponent` means.

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala (hunk 2 above)
Comment (EronWright) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: I agree that it is not essential to expose the utility methods provided by the companion object.

**Comment by jkbradley** [comment]
Thanks for looking--updated.

**Comment by jkbradley** [comment]
Jenkins test this please

**Comment by jkbradley** [comment]
Jenkins test this please

**Comment by jkbradley** [comment]
That was a really strange failure if you look at it.... Will keep an eye on it.
Jenkins test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by jkbradley** [comment]
test this please

**Comment by mengxr** [comment]
LGTM. Merged into master and branch-1.5. Thanks!

**Pull request closed without merging**